src/t_stream.c

```diff
@@ -754,93 +769,53 @@ int64_t streamTrim(stream *s, streamAddTrimArgs *args) {
          * We have to run entry after entry, marking entries as deleted
-         * if they are already not deleted. */
-        while (p) {
-            /* We keep a copy of p (which point to flags part) in order to
-             * update it after (and if) we actually remove the entry */
-            unsigned char *pcopy = p;
-
-            int64_t flags = lpGetInteger(p);
-            p = lpNext(lp, p); /* Skip flags. */
-            int64_t to_skip;
-
-            int64_t ms_delta = lpGetInteger(p);
-            p = lpNext(lp, p); /* Skip ID ms delta */
-            int64_t seq_delta = lpGetInteger(p);
-            p = lpNext(lp, p); /* Skip ID seq delta */
-
-            streamID currid = {0}; /* For MINID */
-            if (trim_strategy == TRIM_STRATEGY_MINID) {
-                currid.ms = master_id.ms + ms_delta;
-                currid.seq = master_id.seq + seq_delta;
-            }
-
-            int stop;
-            if (trim_strategy == TRIM_STRATEGY_MAXLEN) {
-                stop = s->length <= maxlen;
-            } else {
-                /* Following IDs will definitely be greater because the rax
-                 * tree is sorted, no point of continuing. */
-                stop = streamCompareID(&currid, id) >= 0;
-            }
-            if (stop)
-                break;
-
-            if (flags & STREAM_ITEM_FLAG_SAMEFIELDS) {
-                to_skip = master_fields_count;
-            } else {
-                to_skip = lpGetInteger(p); /* Get num-fields. */
-                p = lpNext(lp,p); /* Skip num-fields. */
-                to_skip *= 2; /* Fields and values. */
-            }
-
-            while(to_skip--) p = lpNext(lp,p); /* Skip the whole entry. */
-            p = lpNext(lp,p); /* Skip the final lp-count field. */
-
-            /* Mark the entry as deleted. */
-            if (!(flags & STREAM_ITEM_FLAG_DELETED)) {
-                intptr_t delta = p - lp;
-                flags |= STREAM_ITEM_FLAG_DELETED;
-                lp = lpReplaceInteger(lp, &pcopy, flags);
-                deleted_from_lp++;
-                s->length--;
-                p = lp + delta;
-            }
-        }
-        deleted += deleted_from_lp;
-
+         * if they are already not deleted and perform a garbage collection
+         * if necessary. */
+        struct trimContext context = {
+            .perform_garbage_collection = 0,
+            .processed_all_trimmed_entries = 0,
+            .master_fields = (uint32_t)master_fields_count,
+            .s = s,
+            .id = id,
+            .master_id = &master_id,
+            .maxlen = maxlen,
+            .trim_strategy = trim_strategy,
+            .num_entries = entries,
+            .num_deleted = num_deleted,
+            .num_deleted_by_trim = 0
+        };
+        lp = lpDeleteRanges(lp, p, listpackGetNextDeletedRangeInXTrim, (void *)&context);
+        deleted += context.num_deleted_by_trim;
```

Comment:
can we record the start and tail that needs to be deleted, and clean this part memory when we need a GC?